Add a prow plugin for PR authors to hold PR merges

stevekuznetsov opened the pull request:
Pull request authors need a means by which they can communicate to the
submit queue that their pull request should not be considered for
merges. This plugin adds that functionality with the `/hold` command.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/area prow
/area mungegithub

/cc @fejta @spxtr

fejta commented on prow/plugins/hold/hold.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func init() {
+	plugins.RegisterIssueCommentHandler(pluginName, handleIssueComment)
```
This should also work with a review comment, in addition to an issue comment

See https://github.com/kubernetes/test-infra/blob/master/prow/plugins/lgtm/lgtm.go#L52

stevekuznetsov replied:
Good point! I'll add that.

stevekuznetsov replied:
I'll also add support for it being in the review body since we do that in `/lgtm`

fejta commented on prow/plugins/hold/hold.go:
```diff
@@ -0,0 +1,116 @@
+		return nil
+	}
+
+	if e.commentAuthor != e.prAuthor {
```
I prefer to always start with an assumption that our community is composed of collaborative people with good intentions, and make system as flexible and collaborative as possible.

If we find an individual acting poorly my instinct would be to start by attempting to inform that person about our community's social norms. 

If we discover that our assumption is wrong and our community is riddled with trolls then that is a great time to lock down useful functionality.

stevekuznetsov replied:
Ah, ok! I misunderstood the requirements you put out on Slack! I'll cut this bit out -- it will make the flow simpler as well!

stevekuznetsov commented:
@fejta addressed comments ... fighting with an error I do not understand from the compiler... 

```
prow/github/util_test.go:58: cannot use fc (type *fakegithub.FakeClient) as type labelLister in argument to HasLabel:
	*fakegithub.FakeClient does not implement labelLister (wrong type for GetIssueLabels method)
		have GetIssueLabels(string, string, int) ([]github.Label, error)
		want GetIssueLabels(string, string, int) ([]Label, error)
```

As far as I can tell those types are identical (`Label` and `github.Label`). 

Also, utility packages suck so I would like feedback on how i pulled that method out from `lgtm`

stevekuznetsov commented:
So that was actually an import cycle I created but the compilation in `bazel` didn't show that. Weird.

fejta commented:
Can you resolve the merge conflict? I would close/open to retest travis but looks like you'll need to repush anyway

stevekuznetsov commented:
@fejta rebased & squashed

fejta commented on prow/github/util/util.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+// HasLabel checks if a label is applied to a pr.
+func HasLabel(c labelLister, org, repo string, num int, label string) (bool, error) {
```
@spxtr want to review this expansion of the core prow api?

stevekuznetsov replied:
Much appreciated, I was not super clear on how to do this since `util` packages always kind of suck

spxtr replied:
I agree, `util` packages suck. I think for this one you can just do a `GetIssue` followed by [`Issue.HasLabel`](https://godoc.org/github.com/kubernetes/test-infra/prow/github#Issue.HasLabel). No need for the util package, and it will only add a couple lines elsewhere.

fejta commented:
Uh, also don't forget to import hold into cmd/hook and bump!

spxtr commented on prow/plugins/hold/hold.go:
Check these before using any API tokens.

stevekuznetsov replied:
Done

stevekuznetsov commented:
@spxtr updated -- there was no existing `GetIssue()` call so I kept the current approach -- would you rather I added the call to the client?

spxtr commented on prow/plugins/hold/hold.go:
```diff
@@ -0,0 +1,169 @@
+		return nil
+	}
+
+	var (
```
nit: this block isn't really necessary now, just put the org and such in the event.

stevekuznetsov replied:
They're used both in the event and in the `hasLabel` function for the event, did you mean I should just write them out long-form in both situations?

spxtr replied:
Oh, I missed the `hasLabel` usage. Up to you, then. I don't mind either way.

stevekuznetsov replied:
Less pressure on that overworked CI bot if we don't push another change :)

spxtr commented on prow/plugins/hold/hold.go:
nit: this block isn't really necessary now, just put the org and such in the event.

xiangpengzhao commented:
It would be good to document this in [Pull Request Process](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md) :)

stevekuznetsov commented:
@spxtr @fejta rebased to get the correct `hook` bump and squashed the `var()` blocks while I was at it

kargakis commented:
lgtm
